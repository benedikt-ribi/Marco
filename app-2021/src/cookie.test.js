/**
 * @jest-environment jsdom
 */

import { describe, expect, test, beforeEach } from "@jest/globals";

describe("Cookies", () => {
  function clearCookies() {
    const cookies = document.cookie.split(";");
    const expiryDate = new Date(0).toUTCString();

    for (let i = 0; i < cookies.length; i++) {
      const cookie = cookies[i];
      const [key] = cookie.trim().split("=");
      document.cookie = `${key}=; expires=${expiryDate}`;
    }
  }

  function getCookieMap() {
    const cookie = document.cookie.split(";").map(s => {
      const [key, value] = s.trim().split("=");
      return { key, value };
    });
    return cookie;
  }

  function getCookieObject() {
    const cookie = {};

    for (const pair of document.cookie.split(";")) {
      const [key, value] = pair.trim().split("=");
      cookie[key] = value;
    }
    return cookie;
  }

  beforeEach(() => {
    clearCookies();
  });

  test("get key/value pairs", () => {
    document.cookie = "lastUsername=John";
    document.cookie = "expertMode=true";

    expect(getCookieMap()).toMatchSnapshot();
  });

  test("get values in object", () => {
    document.cookie = "lastUsername=John";
    document.cookie = "expertMode=true";

    expect(getCookieObject()).toMatchSnapshot();
  });

  test("store and load JSON", () => {
    const person = {
      first: "Bob",
      last: "Hoffman",
      age: 34,
      company: {
        name: "Apple",
      },
    };

    document.cookie = "person=" + JSON.stringify(person);

    expect(document.cookie).toMatchSnapshot();

    const loadedPerson = JSON.parse(getCookieObject()["person"]);

    expect(loadedPerson.first).toBe("Bob");
    expect(loadedPerson.last).toBe("Hoffman");
    expect(loadedPerson.age).toBe(34);
    expect(loadedPerson.company.name).toBe("Apple");
  });
});
