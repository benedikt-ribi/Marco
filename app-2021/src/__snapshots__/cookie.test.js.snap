// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cookies get key/value pairs 1`] = `
Array [
  Object {
    "key": "lastUsername",
    "value": "John",
  },
  Object {
    "key": "expertMode",
    "value": "true",
  },
]
`;

exports[`Cookies get values in object 1`] = `
Object {
  "expertMode": "true",
  "lastUsername": "John",
}
`;

exports[`Cookies get values in object 2`] = `"person={\\"first\\":\\"Bob\\",\\"last\\":\\"Hoffman\\",\\"age\\":34,\\"company\\":{\\"name\\":\\"Apple\\"}}"`;

exports[`Cookies store and load JSON 1`] = `"person={\\"first\\":\\"Bob\\",\\"last\\":\\"Hoffman\\",\\"age\\":34,\\"company\\":{\\"name\\":\\"Apple\\"}}"`;
